### PR TITLE
plan: support caching prepare plan for query on hash partition tables

### LINKED ISSUE #11945: Query results for hash partition table may fail when prepare-plan-cache is open
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
Using hash partition create TPC-C test table, turning on prepare plan cache and run test. 
```
bmsql_district | CREATE TABLE `bmsql_district` (
  `d_w_id` int(11) NOT NULL,
  `d_id` int(11) NOT NULL,
  `d_ytd` decimal(12,2) DEFAULT NULL,
  `d_tax` decimal(4,4) DEFAULT NULL,
  `d_next_o_id` int(11) DEFAULT NULL,
  `d_name` varchar(10) DEFAULT NULL,
  `d_street_1` varchar(20) DEFAULT NULL,
  `d_street_2` varchar(20) DEFAULT NULL,
  `d_city` varchar(20) DEFAULT NULL,
  `d_state` char(2) DEFAULT NULL,
  `d_zip` char(9) DEFAULT NULL,
  PRIMARY KEY (`d_w_id`,`d_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY HASH( `d_w_id` + `d_id` )
PARTITIONS 32 |
```

2. What did you expect to see?
using MySQL client run query:

```
mysql> SELECT d_name, d_street_1, d_street_2, d_city,        d_state, d_zip     FROM bmsql_district     WHERE d_w_id = 871 AND d_id =9;
+--------+------------+----------------+-----------------+---------+-----------+
| d_name | d_street_1 | d_street_2     | d_city          | d_state | d_zip     |
+--------+------------+----------------+-----------------+---------+-----------+
| iXeA9e | krwHghjYth | xJSWhFBsX2VLgf | wQdm889bBfmuygz | FP      | 351911111 |
+--------+------------+----------------+-----------------+---------+-----------+
1 row in set (0.01 sec)
```

3. What did you see instead?
Open prepare plan cache:
```
04:39:12,888 [Thread-12] FATAL  jTPCCTerminal : District for W_ID=871 D_ID=9 not foundMB / 466MB
java.lang.Exception: District for W_ID=871 D_ID=9 not found
	at jTPCCTData.executePayment(jTPCCTData.java:776)
	at jTPCCTData.execute(jTPCCTData.java:99)
	at jTPCCTerminal.executeTransactions(jTPCCTerminal.java:160)
	at jTPCCTerminal.run(jTPCCTerminal.java:88)
	at java.lang.Thread.run(Thread.java:748)
```
Close prepare plan cache

```
Term-00, Running Average tpmTOTAL: 8468.77    Current tpmTOTAL: 563160    Memory Usage: 169MB / 365MB
05:57:50,040 [Thread-49] INFO   jTPCC : Term-00,
05:57:50,040 [Thread-49] INFO   jTPCC : Term-00,
05:57:50,040 [Thread-49] INFO   jTPCC : Term-00, Measured tpmC (NewOrders) = 3792.73
05:57:50,040 [Thread-49] INFO   jTPCC : Term-00, Measured tpmTOTAL = 8468.75
05:57:50,040 [Thread-49] INFO   jTPCC : Term-00, Session Start     = 2019-08-30 05:47:47
05:57:50,040 [Thread-49] INFO   jTPCC : Term-00, Session End       = 2019-08-30 05:57:50
05:57:50,040 [Thread-49] INFO   jTPCC : Term-00, Transaction Count = 84994

```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                              |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-alpha-137-gcc07b1108
Git Commit Hash: cc07b110825eac0528b06738a5d6f9a2d970db88
Git Branch: master
UTC Build Time: 2019-08-27 06:10:32
GoVersion: go version go1.12 linux/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
fix #11945
Prepare plan cache will cache physical table id for partition. Partition id changing as query conditions change. So, the physical table id may also change. Query on partition table will not return correct result if prepare plan cache is on.

### What is changed and how it works?
For hash partition, recalculate physical table id by calculate partition expression.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

 - None

Side effects

 - Possible performance regression

Related changes

 - Need to cherry-pick to the release branch

Release note

### CODE REVIEW

File: planner/core/common_plans.go [hunk 1]
```diff
@@ -385,25 +401,69 @@ func (e *Execute) rebuildRange(p Plan) error {
...
 				x.IndexValues[i] = param.Datum
 			}
 		}
+		if x.PartitionInfo != nil {
+			val := x.IndexValues[x.partitionColumnPos].GetInt64()
+			partitionID := val % int64(x.TblInfo.Partition.Num)
+			x.PartitionInfo = &x.TblInfo.Partition.Definitions[partitionID]
+		}
 		return nil
 	case *BatchPointGetPlan:
```
Comment (eurekaka) [inline planner/core/common_plans.go]: Shall `BatchPointGetPlan` be handled as well?

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -376,6 +376,22 @@ func (e *Execute) rebuildRange(p Plan) error {
...
+					if err != nil {
+						return err
+					}
+					pos, err := locateHashPartition(e.ctx, expr[0], pi, ts.Ranges[0].LowVal)
+					if err != nil {
+						return err
+					}
+					pID := pi.Definitions[pos].ID
+					ts.physicalTableID = pID
+				}
```
[the remark this answers is not in the document]
Comment (eurekaka) [reply planner/core/common_plans.go]: Wrong result would be returned for range partitioned table as well?

File: planner/core/common_plans.go (hunk 1 above)
Comment (imtbkcat) [reply planner/core/common_plans.go]: Table partition does not support `BatchPointGet` yet.

File: planner/core/common_plans.go [hunk 3]
```diff
@@ -385,25 +401,69 @@ func (e *Execute) rebuildRange(p Plan) error {
...
+		}
 	case *PointGetPlan:
 		if x.HandleParam != nil {
 			x.Handle, err = x.HandleParam.Datum.ToInt64(sc)
 			if err != nil {
 				return err
 			}
+			if x.PartitionInfo != nil {
+				num := x.TblInfo.Partition.Num
+				pos := x.Handle % int64(num)
```
[the remark this answers is not in the document]
Comment (imtbkcat) [reply planner/core/common_plans.go]: Currently not, i think it still need an `abs` function.

File: planner/core/common_plans.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (imtbkcat) [reply planner/core/common_plans.go]: If it is not a single point plan, it will be a full table scan.

**Comment by imtbkcat** [comment]
PTAL @tiancaiamao

File: planner/core/common_plans.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (eurekaka) [reply planner/core/common_plans.go]: For range partitioned table, shall the partition ID of the cached plan be updated as well?

**Comment by tiancaiamao** [comment]
Please address comment, and add some test code.

**Comment by imtbkcat** [comment]
test has been added, PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

**Comment by imtbkcat** [comment]
PTAL @eurekaka @winoros

**Pull request merged**